### ISSUE TITLE
permanent popups

### ISSUE THREAD

**Issue by methodds** [python-visualization/folium#210]
Hi, I have a question regarding popups for markers. Is it possible to use the markers as a kind of label, such that the text for popups is already visible without having to click on the marker? I would like to use this functionality to label some cities on a map.

**Comment by andrewgiessel** [comment]
Great idea. I bet it's possible. Folium uses leaflet under the hood, so I'd
check out the leaflet API (or plugins) and if it's there we can implement
it. Link us back with anything you find!

**Comment by methodds** [comment]
Thanks for the fast response! Apparently there is a simple plugin for this functionality in leaflet:
https://github.com/Leaflet/Leaflet.label
Can you work with this?

**Comment by themiurgo** [comment]
@methodds @raidoz Would any of you be interested in implementing it? It should be fairly easy to contribute it as a new class heavily similar to Popup.

**Comment by cgarciaf** [comment]
Can you pls show some code line ass un example i need to use this feature....

**Comment by ocefpaf** [comment]
@cgarciaf take a look at the plugins folder in `folium`.

**Comment by cgarciaf** [comment]
Could you pls drive me a bit more i looked into plugin but i am lost...

**Comment by ocefpaf** [comment]
> Could you pls drive me a bit more i looked into plugin but i am lost...

Sorry, but not right now. Folium is maintained by  the efforts of brave volunteers with demanding day jobs :crying_cat_face: 

As soon as I get some free time I will put a skeleton PR together and ping you.

**Comment by cgarciaf** [comment]
You all have me recognition that before hand. Looking fwrd that skeleton, i am lost on the usage.. Thanks
C

**Comment by pcohen89** [comment]
@cgarciaf did anyone ever create some mechanism to add static text (or always open popup text) to folium maps?

**Comment by dunguyen84** [comment]
Dear all,

https://github.com/Leaflet/Leaflet.label   was promising, and after figuring out how to plot the polygon (somehow the methods and the tutorial are not exactly matching anymore), I would love to add the mouse over text poping up. 
I saw the Leaflet.label package but do not know how to integrate in my code. Nor could I add a popup as for folium.Polyline or folium.CircleMarker
I can share my bit of code if someone is willing to help me.

Many thanks in advance, and thank you for folium

**Comment by ocefpaf** [comment]
> I can share my bit of code if someone is willing to help me.

If you can try to do that in the form of a PR I can take a look.

**Comment by jwhendy** [comment]
Man, what a saga on this! I was going to open up a new issue, but *think* this is what I want. I found this example, which is from Leaflet ([source page](http://leafletjs.com/examples/quick-start/) and [standalone example](http://leafletjs.com/examples/quick-start/example-popups.html)), but don't know how to implement in `folium`. Their example has the popup on by default on page load:

![2017-11-23_135954](https://user-images.githubusercontent.com/379357/33187340-bcb242de-d056-11e7-977f-9b5640d85b91.png)

Per the comments on `Leaflet.label`, it says it's now deprecated in favor of `L.tooltip`. I don't see that implemented in `folium` unless it's the same as `folium.map.Marker(..., tooltip='foo')`.

Here's the popup they add above:

```
L.marker([51.5, -0.09]).addTo(mymap)
    .bindPopup("<b>Hello world!</b><br />I am a popup.").openPopup();
```

Does this exist in `folium`?

**Comment by jwhendy** [comment]
Ok, I have approximately *zero* clue if this is even halfways correct. I have no `js` experience whatsoever. I happened to find an instance of `openPopup()` in [search.py](https://github.com/python-visualization/folium/blob/e4663545f243fec09b901936112f32bec6c860bb/folium/plugins/search.py#L68), so I figured it was possible.

I started hunted around in `map.py` for where the popups get added and the [`tooltip` code](https://github.com/python-visualization/folium/blob/master/folium/map.py#L238) in `Marker` caught my eye:

```
{% if this.tooltip %}.bindTooltip("{{this.tooltip.__str__()}}"){% endif %}
```

I just guessed and was able to [get something implemented](https://github.com/jwhendy/folium/blob/open-popup/folium/map.py#L252) in `Popup`:

```
class Popup(Element):
    """Create a Popup instance that can be linked to a Layer.
    Parameters
    ----------

        ### unchanged code removed ###

        self.default_open = default_open

        self._template = Template(u"""
            var {{this.get_name()}} = L.popup({maxWidth: '{{this.max_width}}'});
            {% for name, element in this.html._children.items() %}
                var {{name}} = $('{{element.render(**kwargs).replace('\\n',' ')}}')[0];
                {{this.get_name()}}.setContent({{name}});
            {% endfor %}
            {{this._parent.get_name()}}.bindPopup({{this.get_name()}})
            {% if this.default_open %}.openPopup(){% endif %};
            {% for name, element in this.script._children.items() %}
                {{element.render()}}
            {% endfor %}
        """)  # noqa
```

Let me know if this is sane at all. It works for me; defaults off with no argument (opens on click), defaults on if you pass `default_open=True`. Here's a test:

```
import folium

m = folium.Map(location=[45, -93],
               zoom_start=11)
folium.Marker([45, -93],
              popup=folium.Popup('test popup',
                                 default_open=True)).add_to(m)

m
```

**Comment by jwhendy** [comment]
Update: the above turned out to only work with one popup if multiple were specified. I [found out](https://stackoverflow.com/questions/9047931/leaflet-js-open-all-popup-bubbles-on-page-load) that there's a `leaflet` argument [called `autoClose`](http://leafletjs.com/reference-1.0.2.html#popup-autoclose), which is needed to prevent previous popups from closing when another is opened.

Also probably hokey, but this worked:
```
var {{this.get_name()}} = L.popup({maxWidth: '{{this.max_width}}'
{% if this.default_open %}, autoClose: false{% endif %}});
```

Using two in a test, they're both open on load. Interestingly, if you click one twice, the first click turns it off and the second click turns it back on while auto-closing the second. So, not sure if that's a bug in `autoClose`, if it's intended, or if it only works on page load and not subsequent clicks?

**Comment by jwhendy** [comment]
Looks like there are some other options around keeping popups open. See [this issue at leaflet](https://github.com/Leaflet/Leaflet/issues/5719). 

[One](http://leafletjs.com/reference-1.0.2.html#map-closepopuponclick), `closePopupOnClick` would need to get passed to `L.map()` itself and never close the popups once open (my interpretation, at least).

[The other](http://leafletjs.com/reference-1.0.2.html#popup-closeonclick), would get passed to `L.popup`.  I think this makes more sense with the `autoClose` option, though users might like truly persistent labels on markers as well.

**Comment by jwhendy** [comment]
Totally understood. I'll submit one anyway, though there should likely be some input from the community. I'll submit what I have and create some discussion items that came up in my research of available options. Good luck with the end of the year wrapup!

**Comment by Conengmo** [comment]
Merged in #778

**Comment by sskim-aracnt** [comment]
> Can you pls show some code line as an example i need to use this feature....
try this

```
import folium

m = folium.Map(location=[45, -93],
               zoom_start=11)
folium.Marker([45, -93],
              **popup=folium.Popup('test popup',
                                 show=True)**).add_to(m)
```
